### ISSUE TITLE
Stack requirement reduction

### ISSUE THREAD

**Issue by rkuris** [ava-labs/firewood#538]
When working with the benchmarks, it was observed that we use a LOT of stack space. One culprit was the IO subsystem.

We shouldn't be writing things recursively. I think the entire trie is written in a recursive way. Probably better to queue up all the writes and do them in disk position order or something.

_Originally posted by @rkuris in https://github.com/ava-labs/firewood/pull/528#discussion_r1493068905_

**Comment by rkuris** [comment]
Really tricky to get a stack when the stack overflows.
I was able to get one though. It overflows the stack from the DiskBuffer thread, with this repeating stack, thousands of levels deep:
```
frame #23670: 0x0000aaaaaacd4ca4 process-server`_$LT$futures_util..future..future..shared..Shared$LT$Fut$GT$$u20$as$u20$core..future..future..Future$GT$::poll::hc58fde6306bccf5d [inlined] growthring::wal::WalFilePool$LT$F$C$S$GT$::write::_$u7b$$u7b$closure$u7d$$u7d$::hdbdb47f7868e5ce4((null)=0x0000fffff6840b10) at wal.rs:425:22
    frame #23671: 0x0000aaaaaacd4ca4 process-server`_$LT$futures_util..future..future..shared..Shared$LT$Fut$GT$$u20$as$u20$core..future..future..Future$GT$::poll::hc58fde6306bccf5d(self=Pin<&mut futures_util::future::future::shared::Shared<growthring::wal::{impl#10}::write::{async_block_env#2}<growthring::WalFileImpl, growthring::WalStoreImpl>>> @ x20, cx=<unavailable>) at shared.rs:322:31
    frame #23672: 0x0000aaaaaacd4cb8 process-server`_$LT$futures_util..future..future..shared..Shared$LT$Fut$GT$$u20$as$u20$core..future..future..Future$GT$::poll::hc58fde6306bccf5d [inlined] _$LT$core..pin..Pin$LT$P$GT$$u20$as$u20$core..future..future..Future$GT$::poll::h522bcd17dcce18a2(self=<unavailable>, cx=0x0000fffff6840d00) at future.rs:125:9
    frame #23673: 0x0000aaaaaacd4ca4 process-server`_$LT$futures_util..future..future..shared..Shared$LT$Fut$GT$$u20$as$u20$core..future..future..Future$GT$::poll::hc58fde6306bccf5d [inlined] growthring::wal::WalFilePool$LT$F$C$S$GT$::write::_$u7b$$u7b$closure$u7d$$u7d$::hdbdb47f7868e5ce4((null)=0x0000fffff6840d00) at wal.rs:425:22
    frame #23674: 0x0000aaaaaacd4ca4 process-server`_$LT$futures_util..future..future..shared..Shared$LT$Fut$GT$$u20$as$u20$core..future..future..Future$GT$::poll::hc58fde6306bccf5d(self=Pin<&mut futures_util::future::future::shared::Shared<growthring::wal::{impl#10}::write::{async_block_env#2}<growthring::WalFileImpl, growthring::WalStoreImpl>>> @ x20, cx=<unavailable>) at shared.rs:322:31
```

TLDR is that wal.rs is calling `await` of the previous future, which has another call to await for the next previous future, and so on.

Attempting to run these futures in parallel was fruitless -- the get_file method uses a RefCell, and borrows it mutably, and therefore cannot be called concurrently from multiple threads without a panic.

This wal system needs rearchitecting for simplicity to eliminate this loop.

**Comment by rkuris** [comment]
Solved with new design
